architecture-compose/src/androidLibMain/kotlin/com/splendo/kaluga/architecture/compose/navigation/ModalBottomSheetNavigator.kt

```diff
@@ -1,86 +1,124 @@
+            }
         }
+        sheetContentRouteController.navigate(newRoute)
     }
 
-    fun hide() {
+    override fun back(): Boolean = navHostController.popBackStack() || kotlin.run {
+        close()
+        true
+    }
+
+    override fun close() {
         coroutineScope.launch {
-            modalBottomSheetState.hide()
+            sheetState.animateTo(ModalBottomSheetValue.Hidden)
         }
     }
+}
 
-    @Composable
-    override fun SetupNavHost(builder: NavGraphBuilder.() -> Unit) {
-        val hiddenRoute = CloseBottomSheetNavigationAction.route()
-        SetupNavHost(
-            startDestination = hiddenRoute
-        ) {
-            composable(hiddenRoute) {
-                LaunchedEffect(Unit) {
-                    modalBottomSheetState.hide()
-                }
-            }
-            builder()
-        }
+/** Navigator for [ModalBottomSheetLayout]. */
+class ModalBottomSheetNavigator<A : NavigationAction<*>>(
+    internal val contentRouteController: RouteController,
+    internal val sheetContentRouteController: BottomSheetRouteController,
+    private val navigationMapper: (A) -> BottomSheetRoute
+) : Navigator<A> {
 
-        if (!modalBottomSheetState.isVisible) {
-            syncHiddenState()
+    constructor(
+        contentNavHostController: NavHostController,
+        sheetContentNavHostController: NavHostController,
+        sheetState: ModalBottomSheetState,
+        coroutineScope: CoroutineScope,
+        navigationMapper: (A) -> BottomSheetRoute
+    ) : this(
+        NavHostRouteController(contentNavHostController),
+        BottomSheetRouteController(sheetContentNavHostController, sheetState, coroutineScope),
+        navigationMapper
+    )
+
+    override fun navigate(action: A) {
+        when (val bottomSheetRoute = navigationMapper(action)) {
+            is BottomSheetRoute.SheetContent -> sheetContentRouteController.navigate(
+                bottomSheetRoute.route
+            )
+            is BottomSheetRoute.Content -> contentRouteController.navigate(bottomSheetRoute.route)
         }
     }
+}
+
+typealias BottomSheetContentBuilder = NavGraphBuilder.(contentNavHostController: NavHostController, sheetContentNavHostController: NavHostController, sheetState: ModalBottomSheetState) -> Unit
 
-    private fun syncHiddenState() {
-        navigate(CloseBottomSheetNavigationAction)
+@Composable
+fun <A : NavigationAction<*>> NavigatingModalBottomSheetLayout(
+    navigator: ModalBottomSheetNavigator<A>,
+    sheetContent: BottomSheetContentBuilder,
+    contentRoot: @Composable BottomSheetContentBuilder,
+    content: BottomSheetContentBuilder,
+) = ModalBottomSheetLayout(
+    sheetContent = {
+        if (navigator.sheetContentRouteController.sheetState.isVisible) {
+            HardwareBackButtonNavigation(onBackButtonClickHandler = navigator.sheetContentRouteController::close)
+        }
+        Box(Modifier.defaultMinSize(minHeight = 1.dp)) {
```

Comment:
Yeah, you get a crash otherwise:
https://stackoverflow.com/questions/68623965/jetpack-compose-modalbottomsheetlayout-throws-java-lang-illegalargumentexception